### PR TITLE
implement fetching iot aggregates

### LINKED ISSUE #42: Support reading aggregates from SAP IoT
### Is your feature request related to a problem? Please describe.
Currently project 'Sailor' only supports downloading indicator raw data, leaving all aggregation to the local user. While that is supported through helper functions like `TimeseriesDataset.aggregate`, but in order to reduce local workload and transfer volume it would be good to support also downloading the pre-aggregated data from SAP IoT.

### Describe the solution you'd like
A new function analog to `get_indicator_data` that returns aggregates (and lets the user chose the aggregation functions and aggregation interval).

### PR DESCRIPTION
**Pull request by atrejut**
# Description

Implemented fetching of aggregates from SAP IoT. These aggregates are retrieved from the IoT Hot Store, and pre-aggregated at the server side, reducing the volume of data transferred. They otherwise follow the existing methods for retrieving raw data from SAP IoT.

Until a bugfix in the SAP IoT systems is deployed to live this will only support a subset of aggregates.

Fixes #42 

# Additional information

The new API falls under the umbrella of the extension APIs that have already been approved for use.

# Checklist

Mark "not applicable" if item on the list does not apply to this pull request.

- [x] I have created/adapted unit tests for new code
  - [ ] not applicable 
- [x] I have added new dependencies as described at the [Contributing](https://sap.github.io/project-sailor/contributing.html#requirements-management) page
  - [ ] not applicable 
- [x] I have made corresponding changes to the documentation (incl. tutorial, etc.)
  - [ ] not applicable 
- [x] I have provided release notes (in description or as comment) if this PR is a new feature OR a change worth mentioning for next release
  - [ ] not applicable 
- [ ] I have obtained API approvals if a new SAP API or endpoint is used
  - [x] not applicable

### CODE REVIEW

File: sailor/sap_iot/fetch_aggregates.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+_ALL_KNOWN_AGGREGATION_FUNCTIONS = ['MIN', 'AVG', 'FIRST', 'MAX', 'SUM', 'LAST', 'COUNT',
+                                    'PERCENT_GOOD', 'TMAX', 'TLAST', 'STDDEV', 'TFIRST', 'TMIN']
+
+
+def _parse_aggregation_interval(aggregation_interval):
+    if aggregation_interval is not None and not isinstance(aggregation_interval, str):
+        total_seconds = aggregation_interval.total_seconds()
+        aggregation_interval = f'PT{total_seconds}S'
+    return aggregation_interval
```
Comment (cloudyday) [inline sailor/sap_iot/fetch_aggregates.py:40]: do we actually want to allow setting the aggregation interval to "None"? 
what you get back is a single data point aggregating the whole duration. this behavior is also not clearly defined, is it? it just happens.
do you see any benefits of offering that?

**Review by cloudyday — COMMENTED** [review]
preliminary review.

most important remark: this only works with furl>=2.1.3 because of the `;` parsing in the "next link" of the pagination.
You need to add this into the requirements.in (with that done we're good)

File: sailor/sap_iot/fetch_aggregates.py (hunk 1 above)
Comment (atrejut) [reply sailor/sap_iot/fetch_aggregates.py:40]: as discussed, let's leave it possible to use None

File: sailor/assetcentral/equipment.py [hunk 2]
```diff
@@ -380,14 +377,56 @@ def get_indicator_data(self, start: Union[str, pd.Timestamp, datetime.timestamp,
...
             indicator_set = self.find_common_indicators()
 
         return sap_iot.get_indicator_data(start, end, indicator_set, self)
 
+    def get_indicator_aggregates(
+            self, start: Union[str, pd.Timestamp, datetime.timestamp, datetime.date],
+            end: Union[str, pd.Timestamp, datetime.timestamp, datetime.date],
+            indicator_set: IndicatorSet = None,
+            aggregation_functions: Iterable[str] = ('AVG',),
+            aggregation_interval: Union[str, pd.Timedelta, timedelta] = 'PT2M'
```
[the remark this answers is not in the document]
Comment (atrejut) [reply sailor/assetcentral/equipment.py:390]: thx

File: sailor/sap_iot/fetch_aggregates.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        duration = results[0]['properties']['duration'] if duration is None else duration
+        df = pd.merge(df, results_df, on=['timestamp', 'equipment_id'], how='outer')
+
+    if df is None:
+        raise RuntimeError("No data could be found for the given parameters.")
+
+    df['timestamp'] = pd.to_datetime(df['timestamp'])
+
+    if aggregation_interval is not None:
+        duration = pd.Timedelta('P' + duration)
```
Comment (cloudyday) [inline sailor/sap_iot/fetch_aggregates.py:151]: the problem with the ISO 8601 duration returned by the endpoint and pd Timedelta as mentioned

File: sailor/sap_iot/fetch_aggregates.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        LOG.debug("Fetching indicator data. Indicator Group ID: %s.", liot_group_id)
+        results = _fetch_aggregates(liot_group_id, template_id, oauth_iot, **params)
+        results_df = _prepare_df(results, aggregated_indicators, liot_group_id, template_id)
+        if results_df.empty:
+            continue
+        duration = results[0]['properties']['duration'] if duration is None else duration
+        df = pd.merge(df, results_df, on=['timestamp', 'equipment_id'], how='outer')
+
+    if df is None:
+        raise RuntimeError("No data could be found for the given parameters.")
```
Comment (cloudyday) [inline sailor/sap_iot/fetch_aggregates.py:146]: This check actually does not work. DF is always set. And when the for loop doesn't return any data, `duration` is None and it runs into an error in line 151 (TypeError: can only concatenate str (not "NoneType") to str).

This also warrants a test for the case that "no data is found".

And we should handle it differently. In other areas when no data is found, we don't throw an error. Instead we throw a warning and return an empty result.

**Pull request merged**